### ISSUE TITLE
Whole Words search in the editor is very slow in large projects

### ISSUE THREAD

**Issue by ondesic** [godotengine/godot#33473]
**3.2.alpha.calinou.af4fd9de9**

Working on a large GDScript project with 25k line count I notice a few things that don't occur on a small project.
Turn on "Whole Words" in the search panel, then begin to type what you need to find.  Godot becomes unresponsive for over 10 minutes.  It doesn't do this in 3.1  Something has changed that bogs down the editor

**Comment by Calinou** [comment]
Does this happen in a smaller project with less lines of code?

**Comment by ondesic** [comment]
in 3.2 the pause is much shorter in smaller projects.
However, in 3.1 this "Whole Words" search problem does not occur even in my large project

**Comment by Ranoller** [comment]
At 21/01/2020 Whole Words stucks editor 40 seconds in a search that have 20 matches without that option and 8 matches with that, in a script with 5223 lines, in a project with 44342 lines of code.
 
Same search in a old plugin of mine writed in gdscript (previous to that functionality in godot) in the entire project (Analize 114 scripts plus 109 tscn files, 80 tres files, 30 json files and 32 wiki files) takes 3 seconds... so there are some problem in that option.

**Comment by Calinou** [comment]
@ondesic @Ranoller Can you reproduce this in earlier Godot 3.2 betas or alphas?

**Comment by Ranoller** [comment]
My version is from 4 days ago... but i reading now code_editor.cpp, I remember that behavior from some time ago but i´m not use this option so much. Bibisect can be troublesome. Indeed, i´m trying to find some in the code. I will report if i see something souspicious.

**Comment by Ranoller** [comment]
I think it is related with the "match count" functionality. It reports bad results. A word that is not substring of others is reported 6 times without "whole words" (correct) and 3 with "Whole words" (bad result, is 6 with and without whole words), same with other searchs... the finded words are correct (You got that words by pressing enter) but the count result is bad (allways less words that the words finded)

**Comment by Ranoller** [comment]
Well, a accesory bug is that **a "whole word is not counted" if is in the end of the line**, i think is because you are counting only words with symbols around, but end of line maybe is not a symbol

I talking about this piece of code (code_editor.cpp lines 307++):

```

	int from_pos = 0;

	while (true) {
		int pos = is_case_sensitive() ? full_text.find(searched, from_pos) : full_text.findn(searched, from_pos);
		if (pos == -1) break;

		if (is_whole_words()) {
			from_pos++; // Making sure we won't hit the same match next time, if we get out via a continue.
			if (pos > 0 && !is_symbol(full_text[pos - 1]))
				continue;
			if (pos + searched.length() < full_text.length() && !is_symbol(full_text[pos + searched.length()]))
				continue;
		}

		results_count++;
		from_pos = pos + searched.length();
	}
}


```

Is possible to fix that using _is_text_char replacing !is_symbol or something like that? (Is hard to me follow string parse code, maybe is not the solution)

**Comment by akien-mga** [comment]
This should probably be refactored to use `RegEx`, it would be a lot more efficient than using `String.find` repeatedly like htis.

**Comment by squished-01** [comment]
Would love it if someone smarter than me could please fix this and cherry pick it into the 3.2 branch.  I'm surprised this hasn't bothered more people, are we all just avoiding "whole words" in large files?
